As electronic vehicles (EVs) become more popular, there is an increasing need for access to charging stations, also known as ports. To that end, many modern apartment buildings have begun retrofitting their parking garages to include shared charging stations. A charging station is shared if it is accessible by anyone in the building.

<img src="charging_station.jpg" alt="EV Charging" width="500" height="auto">

But with increasing demand comes competition for these ports — nothing is more frustrating than coming home to find no charging stations available! In this project, you will use a dataset to help apartment building managers better understand their tenants’ EV charging habits.

The data has been loaded into a PostgreSQL database with a table named `charging_sessions` with the following columns:

## charging_sessions

| Column | Definition | Data type |
|-|-|-|
|`garage_id`| Identifier for the garage/building|`VARCHAR`|
|`user_id` | Identifier for the individual user|`VARCHAR`|
|`user_type`|Indicating whether the station is `Shared` or `Private`| `VARCHAR` |
|`start_plugin`|The date and time the session started |`DATETIME`|
|`start_plugin_hour`|The hour (in military time) that the session started | `NUMERIC`|
|`end_plugout`|The date and time the session ended | `DATETIME` |
|`end_plugout_hour`|The hour (in military time) that the session ended | `NUMERIC`|
|`duration_hours`| The length of the session, in hours|`NUMERIC`|
|`el_kwh`| Amount of electricity used (in Kilowatt hours)|`NUMERIC`|
|`month_plugin`| The month that the session started |`VARCHAR`|
|`weekdays_plugin`| The day of the week that the session started|`VARCHAR`|

Let’s get started!

#### Sources
- **Data**: [CC BY 4.0](https://creativecommons.org/licenses/by/4.0), via [Kaggle](https://www.kaggle.com/datasets/anshtanwar/residential-ev-chargingfrom-apartment-buildings),
- **Image**: Julian Herzog, [CC BY 4.0](https://creativecommons.org/licenses/by/4.0), via Wikimedia Commons

Problem Statement:

As the popularity of electric vehicles (EVs) continues to grow, the demand for accessible charging stations in apartment buildings has increased. This rising demand can lead to competition for available charging ports, causing frustration among tenants. To address this issue, it is essential to understand the charging habits of tenants and the utilization patterns of the existing charging infrastructure. This analysis aims to provide apartment building managers with actionable insights into EV charging behavior, identify peak usage times, compare the usage of shared versus private stations, and highlight any patterns that could inform better management and expansion of charging facilities. By answering these key questions, we can help optimize the availability and efficiency of EV charging stations, ensuring a better experience for all residents.Let us start the analysis to understand EV charging habits of tenants in apartment buildings. 

Let us start the Analysis......

In [3]:
-- Return a tibble of the table charging_sessions to take a look

SELECT *
FROM charging_sessions
LIMIT 10;

garage_id user_id user_type  ... duration_hours  month_plugin weekdays_plugin
0      AdO3  AdO3-4   Private  ...       0.050000           Dec          Friday
1      AdO3  AdO3-4   Private  ...       0.136667           Dec          Friday
2      AdO3  AdO3-4   Private  ...       8.216389           Dec          Friday
3      AdO3  AdO3-2   Private  ...      24.419722           Dec        Saturday
4      AdO3  AdO3-2   Private  ...       0.970556           Dec          Monday
5      AdO3  AdO3-2   Private  ...      18.077778           Dec          Monday
6      AdO3  AdO3-2   Private  ...      21.720833           Dec         Tuesday
7      AdO3  AdO3-4   Private  ...       6.188056           Dec       Wednesday
8      AdO3  AdO3-2   Private  ...       2.371111           Dec       Wednesday
9      AdO3  AdO3-2   Private  ...       4.856111           Dec        Saturday

[10 rows x 11 columns]

## 1. What are the peak hours for EV charging? 

In [10]:
-- Identifying peak hours helps in managing and possibly expanding charging infrastructure to accommodate high demand times.
SELECT start_plugin_hour, COUNT(*) AS total_sessions
FROM charging_sessions
GROUP BY start_plugin_hour
ORDER BY total_sessions DESC;
/* Looking at the total number of sessions by hour (0 to 23), most people try to charge between 15 (3 PM) to 19 (7 PM). */


start_plugin_hour  total_sessions
0                  16             966
1                  15             787
2                  19             676
3                  17             608
4                  18             584
5                  20             583
6                  21             463
7                  14             408
8                  22             374
9                  13             294
10                 12             202
11                 23             199
12                 11             173
13                 10             116
14                  0             102
15                  8              81
16                  7              79
17                  9              66
18                  1              40
19                  2              25
20                  6              19
21                  3              17
22                  5               9
23                  4               7

## 2. What is the average duration of charging sessions?

In [18]:
-- Understanding the average duration can help in optimizing charging schedules and informing users about expected charging times.
SELECT AVG(duration_hours) AS avg_duration
FROM charging_sessions;
/* Average Duration of charging time is 11.5 Hours. */

user_type  avg_duration
0    Shared      6.535401
1   Private     12.784593

## 3. How does the usage of shared vs. private stations compare along with number of unique users in each station ?

In [16]:
-- This comparison helps in determining if shared stations are sufficient or if more private stations are needed. Also, Number of Unique users based on station. 
SELECT user_type, 
       COUNT(*) AS total_sessions, 
       SUM(el_kwh) AS total_kwh,
       COUNT(DISTINCT user_id) AS unique_users
FROM charging_sessions
GROUP BY user_type;
/* Looking into this information seems like 

user_type  total_sessions  total_kwh  unique_users
0   Private            5466   61231.37            56
1    Shared            1412   26261.91            52

## 4. What are the most common days for charging sessions?

In [21]:
-- Knowing the most common days can assist in planning maintenance and ensuring availability during high-demand days.
SELECT weekdays_plugin, COUNT(*) AS total_sessions
FROM charging_sessions
GROUP BY weekdays_plugin
ORDER BY total_sessions DESC;
/* Most common weekdays to charge EV vehicles are Thursday, Wednesday and Friday. */

weekdays_plugin  total_sessions
0        Thursday            1072
1       Wednesday            1039
2          Friday            1034
3          Monday             981
4          Sunday             955
5         Tuesday             943
6        Saturday             854

## 5. How does charging activity vary by month?

In [39]:
-- Seasonal trends in charging can help in predicting future needs and planning for infrastructure upgrades.
SELECT month_plugin, COUNT(*) AS total_sessions, SUM(el_kwh) AS total_kwh
FROM charging_sessions
GROUP BY month_plugin
ORDER BY total_sessions DESC;
/* Months during the Fall highlight excessive need for the chaging stations. */

month_plugin  total_sessions  total_kwh
0           Jan            1361   16783.46
1           Dec            1143   13823.54
2           Nov            1107   14091.95
3           Oct             845   10821.10
4           Sep             501    7507.08
5           Aug             362    5626.93
6           Apr             306    3175.91
7           Mar             301    3539.82
8           Jun             291    3819.44
9           May             288    3451.67
10          Jul             238    3323.96
11          Feb             135    1528.42

## 6. Which garages/buildings have the highest usage?

In [26]:
-- Identifying high-usage garages can highlight areas needing more charging stations or better management.
SELECT garage_id, COUNT(*) AS total_sessions, SUM(el_kwh) AS total_kwh
FROM charging_sessions
GROUP BY garage_id
ORDER BY total_sessions DESC;
/* Need to Investigate more about on BI2 garage */

garage_id  total_sessions  total_kwh
0        Bl2            2243   27046.92
1       AsO2             665   10672.16
2       AdO1             586    8372.83
3        UT9             421    6789.82
4       AdO3             379    5239.22
5        UT7             313    2532.65
6        MS1             270    1763.17
7      AsO10             253    3428.27
8        NR1             224    1809.98
9        SR2             221    3768.99
10      MS22             187    2651.19
11      UT15             170    1599.28
12       UT1             166    1856.79
13      AdA6             156     616.69
14       Ris             107    1009.37
15      SR14             105    3791.26
16      AsO8             104    1117.18
17      AsO4             103     776.79
18      AdA1              97     926.59
19      AsO6              39     259.53
20       UT2              28     352.11
21       Bl7              19     865.96
22      MS11              14     169.25
23       SR4               8      77.28

## 7. What is the total electricity consumption for each garage/building?

In [40]:
-- Total electricity consumption data is crucial for energy management and planning utility needs.
SELECT garage_id, SUM(el_kwh) AS total_kwh
FROM charging_sessions
GROUP BY garage_id
ORDER BY total_kwh DESC;
/* Similar to what we have observed before; BI2 garage has the highest total electricity consumption. */

garage_id  total_kwh
0        Bl2   27046.92
1       AsO2   10672.16
2       AdO1    8372.83
3        UT9    6789.82
4       AdO3    5239.22
5       SR14    3791.26
6        SR2    3768.99
7      AsO10    3428.27
8       MS22    2651.19
9        UT7    2532.65
10       UT1    1856.79
11       NR1    1809.98
12       MS1    1763.17
13      UT15    1599.28
14      AsO8    1117.18
15       Ris    1009.37
16      AdA1     926.59
17       Bl7     865.96
18      AsO4     776.79
19      AdA6     616.69
20       UT2     352.11
21      AsO6     259.53
22      MS11     169.25
23       SR4      77.28

## 8. What is the average electricity consumption per session?

In [29]:
-- Average consumption per session helps in understanding the efficiency and energy requirements of the charging sessions.
SELECT AVG(el_kwh) AS avg_kwh
FROM charging_sessions;
/* Average consumption per session is 12.72 kwh */

avg_kwh
0  12.720744

## 9. What are the longest charging sessions?

In [41]:
-- Analyzing the longest sessions can reveal outliers and help in setting policies for fair usage.
SELECT *
FROM charging_sessions
ORDER BY duration_hours DESC;
/* This shows Outliers in the data, as end plugout time is not captured for 33 records */

garage_id   user_id user_type  ... duration_hours  month_plugin weekdays_plugin
0          UT7     UT7-1   Private  ...            NaN           Dec        Thursday
1         MS22    MS22-1   Private  ...            NaN           Dec          Monday
2          SR2  Share-17    Shared  ...            NaN           Dec          Monday
3          SR2  Share-32    Shared  ...            NaN           Dec          Monday
4         MS22    MS22-1   Private  ...            NaN           Dec          Monday
...        ...       ...       ...  ...            ...           ...             ...
6873       UT9     Bl2-7    Shared  ...       0.005000           Sep          Sunday
6874      AsO2    AsO2-1   Private  ...       0.004444           Dec       Wednesday
6875       NR1     NR1-3   Private  ...       0.004167           Oct        Thursday
6876       UT7     UT7-3   Private  ...       0.003611           Oct        Thursday
6877       UT9  Share-12    Shared  ...       0.003333           Sep          Monday

[6878 rows x 11 columns]

## 10. How many sessions start during each hour of the day?

In [43]:
--Understanding hourly session starts helps in identifying patterns and peak times for better management.
SELECT start_plugin_hour, COUNT(*) AS total_sessions
FROM charging_sessions
GROUP BY start_plugin_hour
ORDER BY total_sessions DESC;
/* Most charging stations are occupied around 4 PM, causing potential frustration due to unavailability. */


start_plugin_hour  total_sessions
0                  16             966
1                  15             787
2                  19             676
3                  17             608
4                  18             584
5                  20             583
6                  21             463
7                  14             408
8                  22             374
9                  13             294
10                 12             202
11                 23             199
12                 11             173
13                 10             116
14                  0             102
15                  8              81
16                  7              79
17                  9              66
18                  1              40
19                  2              25
20                  6              19
21                  3              17
22                  5               9
23                  4               7

## 11. What is the distribution of session durations?

In [5]:
-- Distribution analysis of session durations helps in planning and optimizing station availability.
SELECT duration_hours, COUNT(*) AS total_sessions
FROM charging_sessions
GROUP BY duration_hours
ORDER BY duration_hours;
/* This Analysis will be helpful in understanding teh trends when projected in the form of a Histogram */

duration_hours  total_sessions
0           0.003333               1
1           0.003611               1
2           0.004167               1
3           0.004444               1
4           0.005000               1
...              ...             ...
6497      235.020278               1
6498      237.033333               1
6499      239.240278               1
6500      255.029444               1
6501             NaN              34

[6502 rows x 2 columns]

## 12. Are there any significant patterns in charging behavior over the week?

In [46]:
SELECT weekdays_plugin, start_plugin_hour, COUNT(*) AS total_sessions
FROM charging_sessions
GROUP BY weekdays_plugin, start_plugin_hour
ORDER BY weekdays_plugin, start_plugin_hour;


weekdays_plugin  start_plugin_hour  total_sessions
0            Friday                  0              11
1            Friday                  1               4
2            Friday                  2               3
3            Friday                  3               1
4            Friday                  4               2
..              ...                ...             ...
158       Wednesday                 19             105
159       Wednesday                 20             113
160       Wednesday                 21              78
161       Wednesday                 22              66
162       Wednesday                 23              23

[163 rows x 3 columns]

## 13. What are the top users based on electricity consumption?

In [36]:
SELECT user_id, SUM(el_kwh) AS total_kwh
FROM charging_sessions
GROUP BY user_id
ORDER BY total_kwh DESC
LIMIT 10;

user_id  total_kwh
0     Bl2-5    3696.16
1    AdO1-3    3587.44
2     Bl2-2    3288.96
3    AsO6-1    3043.23
4    SR14-2    2943.57
5     SR2-3    2922.53
6  Share-12    2743.50
7     Bl2-3    2737.72
8    AsO2-1    2644.69
9     Bl2-1    2579.38

## 14. How does charging activity differ between weekdays and weekends?

In [38]:
-- Differentiating between weekday and weekend activity helps in understanding how usage patterns shift and planning accordingly.
SELECT CASE 
           WHEN weekdays_plugin IN ('Saturday', 'Sunday') THEN 'Weekend'
           ELSE 'Weekday'
       END AS day_type, COUNT(*) AS total_sessions, SUM(el_kwh) AS total_kwh
FROM charging_sessions
GROUP BY day_type;
/* This doesn't give much Info */

day_type  total_sessions  total_kwh
0  Weekend            1809   24815.10
1  Weekday            5069   62678.18

## 15. What are the number of unique users for garage?

In [3]:
-- unique_users_per_garage
-- This query finds the number of unique users per garage.
SELECT garage_id, COUNT(DISTINCT user_id) AS num_unique_users
FROM charging_sessions
WHERE user_type = 'Shared'
GROUP BY garage_id
ORDER BY num_unique_users DESC;
/* BI2 has the highest number of Unique Users */ 

garage_id  num_unique_users
0       Bl2                18
1      AsO2                17
2       UT9                16
3      AdO3                 3
4       MS1                 2
5       SR2                 2
6      AdA1                 1
7       Ris                 1

## Findings
1. **Peak Hours for EV Charging:** The busiest hours for charging sessions are between 3 PM and 7 PM, with the highest number of sessions starting at 4 PM.
2. **Average Duration of Charging Sessions:** The overall average duration of charging sessions is approximately 11.5 hours. Shared stations have shorter average sessions (around 6.5 hours), while private stations have longer sessions (approximately 12.8 hours).
3. **Usage of Shared vs. Private Stations:** 
- Private stations are used more frequently than shared stations, with 5466 sessions compared to 1412 sessions for shared stations.
- Private stations consume more electricity (61231.37 kWh) compared to shared stations (26261.91 kWh).
- The number of unique users for private stations is slightly higher (56) compared to shared stations (52).
4. **Most Common Days for Charging Sessions:** The highest charging activity occurs on Thursdays, Wednesdays, and Fridays.
5. **Monthly Charging Activity:** The fall months (October to December) show the highest charging activity, with December and January being particularly busy.
6. **High-Usage Garages:** The garage with the highest usage is BI2, with 2243 sessions and the highest total electricity consumption (27046.92 kWh).
7. **Electricity Consumption per Session:** The average electricity consumption per session is 12.72 kWh.
8. **Longest Charging Sessions:** Some charging sessions are significantly long, with certain sessions missing end plug-out times, indicating potential outliers or data recording issues.
9. ** Hourly Charging Patterns:** Most charging sessions start around 4 PM, indicating a peak usage time that may lead to frustration due to unavailability.
10. **Top Users Based on Electricity Consumption:** The top users, such as user Bl2-5, have significantly high electricity consumption, with the top 10 users consuming between 2579.38 kWh and 3696.16 kWh.
11. **Weekday vs. Weekend Charging Activity:** Weekdays see higher charging activity (5069 sessions and 62678.18 kWh) compared to weekends (1809 sessions and 24815.1 kWh).
12. **Unique Users per Garage:** The garage BI2 has the highest number of unique users (18) for shared stations, followed by AsO2 (17) and UT9 (16).

## Recommendations

1. **Optimize Charging Infrastructure:**
- Increase the number of charging stations in high-usage garages, particularly BI2, to alleviate congestion during peak hours.
- Consider expanding shared station availability to better accommodate demand, especially during the afternoon peak hours.

2. **Manage Peak Usage Times:**
- Implement a reservation system to allow tenants to book charging slots, especially during peak hours (3 PM to 7 PM).
- Encourage off-peak charging by providing incentives or discounted rates for charging sessions outside the peak hours.

3. **Enhance Data Collection:**
- Address the data recording issues for end plug-out times to ensure accurate tracking of session durations.
- Regularly monitor and clean the data to identify and rectify any outliers or inconsistencies.

4. **Promote Efficient Charging Habits:**
- Educate tenants on optimal charging times and the benefits of shorter charging sessions to improve station availability.
- Implement policies or guidelines to limit excessively long charging sessions, ensuring fair usage for all residents.

5. **Plan for Future Expansion:**
- Based on the increasing trend of EV adoption and the high demand during certain months, plan for future expansions of the charging infrastructure to accommodate growing needs.
- Evaluate the potential for installing additional charging stations in garages with high unique user counts and electricity consumption.

## Conclusion
The analysis of EV charging habits in apartment buildings reveals significant insights into peak usage times, average session durations, and the comparative usage of shared versus private stations. High-demand periods, particularly in the afternoons and specific fall months, highlight the need for optimized charging infrastructure and better management strategies. By implementing the recommendations, apartment building managers can enhance the availability and efficiency of EV charging stations, ultimately providing a better experience for tenants and supporting the growing adoption of electric vehicles.
